src/util/circular_buffer.c

```diff
@@ -0,0 +1,253 @@
+		// only the thread with the largest offset will succeed
+		// for the above example, W1 will succeed
+		//
+		// [., ., ., ., ., ., A, B, C]
+		//        ^
+		//        W
+		atomic_compare_exchange_strong(&cb->write, &overflow, offset + cb->item_size);
+	}
+
+	// copy item into buffer
+	memcpy(cb->data + offset, item, cb->item_size);
+
+	// report success
+	return 1;
+}
+
+// Reserve a slot within buffer.
+// Returns a pointer to a 'item size' slot within the buffer.
+// This function is thread-safe and lock-free.
+void *CircularBuffer_Reserve(CircularBuffer cb) {
+	RedisModule_Assert(cb != NULL);
+
+	// atomic update buffer item count
+	// an item will be overwritten if buffer is full
+	uint64_t item_count = atomic_fetch_add(&cb->item_count, 1);
+	if (unlikely(item_count >= cb->item_cap)) {
+		cb->item_count = cb->item_cap;
+	}
+
+	// determine current and next write position
+	uint64_t offset = atomic_fetch_add(&cb->write, cb->item_size);
+
+	// check for buffer overflow
+	if (unlikely(cb->data + offset >= cb->end_marker)) {
+		// write need to circle back
+		// [., ., ., ., ., ., A, B, C]
+		//                           ^  ^
+		//                           W0 W1
+		uint64_t overflow = offset + cb->item_size;
+
+		// adjust offset
+		// [., ., ., ., ., ., A, B, C]
+		//  ^  ^
+		//  W0 W1
+		offset -= cb->item_size * cb->item_cap;
+
+		// update write position
+		// multiple threads "competing" to update write position
+		// only the thread with the largest offset will succeed
+		// for the above example, W1 will succeed
+		//
+		// [., ., ., ., ., ., A, B, C]
+		//        ^
+		//        W
+		atomic_compare_exchange_strong(&cb->write, &overflow, offset + cb->item_size);
+	}
+
+	// return slot pointer
+	return cb->data + offset;
+}
+
+// Read oldest item from buffer.
+// This function is not thread-safe.
+// This function pops the oldest item from the buffer.
+void *CircularBuffer_Read(CircularBuffer cb, void *item) {
+	RedisModule_Assert(cb != NULL);
+
+	// make sure there's data to return
+	if (unlikely(CircularBuffer_Empty(cb))) {
+		return NULL;
+	}
+
+	void *read = cb->read;
+
+	// update buffer item count
+	cb->item_count--;
+
+	// copy item from buffer to output
+	if (item != NULL) {
+		memcpy(item, cb->read, cb->item_size);
+	}
+
+	// advance read position
+	// circle back if read reached the end of the buffer
+	cb->read += cb->item_size;
+	if (unlikely(cb->read >= cb->end_marker)) {
+		cb->read = cb->data;
+	}
+
+	// return original read position
+	return read;
+}
+
+// Sets the read pointer to the beginning of the buffer.
+// assuming the buffer looks like this:
+//
+// [., ., ., A, B, C, ., ., .]
+//                    ^
+//                    W
+//
+// CircularBuffer_ResetReader will set 'read' to A
+//
+// [., ., ., A, B, C, ., ., .]
+//           ^        ^
+//           R        W
+//
+void CircularBuffer_ResetReader(CircularBuffer cb) {
+	// compensate for circularity
+	uint64_t write = cb->write;
```

Comment:
shouldn't you do an atomic load operation on write?

Reply:
This is done under write lock, so no need for atomicity (not thread safe).
If we will want to later introduce a lock to the Circular Buffer we may do that, but let's first understand that we want\need that. I don't think so, as we will have a lock in the wrapping struct.

Reply:
Yes